hail/src/main/scala/is/hail/expr/ir/Infer.scala

```diff
@@ -1,89 +1,101 @@
 package is.hail.expr.ir
 
-import is.hail.expr.types.{virtual, _}
+import is.hail.expr.types.physical._
 import is.hail.expr.types.virtual._
+import is.hail.utils._
 
 object Infer {
-  def apply(ir: InferIR): Type = {
+  def apply(ir: InferIR): PType = {
     ir match {
       case If(cond, cnsq, altr) =>
         assert(cond.typ.isOfType(TBoolean()))
         assert(cnsq.typ.isOfType(altr.typ))
-        if (cnsq.typ != altr.typ)
-          cnsq.typ.deepOptional()
+        if (cnsq.pType != altr.pType)
+          PType.canonical(cnsq.typ.deepOptional())
         else
-          cnsq.typ
+          cnsq.pType
       case Let(name, value, body) =>
-        body.typ
+        body.pType
       case ApplyBinaryPrimOp(op, l, r) =>
-        BinaryOp.getReturnType(op, l.typ, r.typ)
+        PType.canonical(BinaryOp.getReturnType(op, l.typ, r.typ)).setRequired(l.pType.required && r.pType.required)
       case ApplyUnaryPrimOp(op, v) =>
-        UnaryOp.getReturnType(op, v.typ)
+        PType.canonical(UnaryOp.getReturnType(op, v.typ)).setRequired(v.pType.required)
       case ApplyComparisonOp(op, l, r) =>
         assert(l.typ isOfType r.typ, s"${l.typ.parsableString()} vs ${r.typ.parsableString()}")
-        TBoolean()
+        PBoolean(l.pType.required && r.pType.required)
       case ArrayRef(a, i) =>
         assert(i.typ.isOfType(TInt32()))
-        -coerce[TArray](a.typ).elementType
+        coerce[PArray](a.pType).elementType.setRequired(a.pType.required && i.pType.required)
       case ArraySort(a, ascending, _) =>
         assert(ascending.typ.isOfType(TBoolean()))
-        a.typ
+        val et = coerce[PArray](a.pType).elementType
+        PArray(et, a.pType.required)
```

Comment:
Is this different than just a.pType?

Reply:
right now, no. But when we have non-canonical ptypes, very potentially. It's hard to know what to build for that future world and what to build assuming all-canonical.